### ISSUE TITLE
PolygonScan support

### ISSUE THREAD

**Issue by tjones0808** [rkalis/truffle-plugin-verify#81]

**Comment by rkalis** [comment]
truffle-plugin-verify currently only works with Etherscan and its derivative platforms (like BscScan or HecoInfo). Etherscan has no platform for Matic (yet) so until then it can't be added to this plugin. Explorers like maticvigil are based on blockscout, so I just opened #82 to implement support for blockscout explorers in general (which would include maticvigil). I can't give any timelines on when (if) that will be implemented though.

**Comment by rkalis** [comment]
Since Etherscan has recently launched PolygonScan, I'm reopening this issue. I implemented Polygon support in the polygon branch (https://github.com/rkalis/truffle-plugin-verify/tree/polygon). But it doesn't look to be working yet, which appears to be an issue on the PolygonScan side.

When I deploy the same contracts (the example `metacoin` directory) to Rinkeby and Polygon, and then run verify, it works for Rinkeby, but not Polygon. Debug output:

Rinkeby:
```
rosco@rmbp ~/r/e/t/t/metacoin (master)> truffle run verify MetaCoin --network rinkeby --debug
DEBUG logging is turned ON
Running truffle-plugin-verify v0.5.8
Verifying MetaCoin
Reading artifact file at /Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/build/contracts/MetaCoin.json
Retrieving constructor parameters from https://api-rinkeby.etherscan.io/api?apiKey=<redacted>&module=account&action=txlist&address=0x9731319B3aFE83E2E43e05894eF63A4864897572&page=1&sort=asc&offset=1
Constructor parameters retrieved: 0x
Reading artifact file at /Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/build/contracts/ConvertLib.json
Sending verify request with POST arguments:
{
  "apikey": "<redacted>",
  "module": "contract",
  "action": "verifysourcecode",
  "contractaddress": "0x9731319B3aFE83E2E43e05894eF63A4864897572",
  "sourceCode": "{\"language\":\"Solidity\",\"sources\":{\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/ConvertLib.sol\":{\"content\":\"// SPDX-License-Identifier: MIT\\npragma solidity >=0.4.25 <0.9.0;\\n\\nlibrary ConvertLib {\\n\\tstruct TestStruct {\\n\\t\\tuint256 name;\\n\\t}\\n\\n\\tfunction convert(uint amount,uint conversionRate) public pure returns (uint convertedAmount)\\n\\t{\\n\\t\\treturn amount * conversionRate;\\n\\t}\\n}\\n\"},\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/MetaCoin.sol\":{\"content\":\"// SPDX-License-Identifier: MIT\\npragma experimental ABIEncoderV2;\\npragma solidity >=0.4.25 <0.9.0;\\n\\nimport \\\"./ConvertLib.sol\\\";\\n\\n// This is just a simple example of a coin-like contract.\\n// It is not standards compatible and cannot be expected to talk to other\\n// coin/token contracts. If you want to create a standards-compliant\\n// token, see: https://github.com/ConsenSys/Tokens. Cheers!\\n\\ncontract MetaCoin {\\n\\tmapping (address => uint) balances;\\n\\n\\tevent Transfer(address indexed _from, address indexed _to, uint256 _value);\\n\\n\\tconstructor() public {\\n\\t\\tbalances[tx.origin] = 10000;\\n\\t}\\n\\n\\tfunction sendCoin(address receiver, uint amount) public returns(bool sufficient) {\\n\\t\\tif (balances[msg.sender] < amount) return false;\\n\\t\\tbalances[msg.sender] -= amount;\\n\\t\\tbalances[receiver] += amount;\\n\\t\\temit Transfer(msg.sender, receiver, amount);\\n\\t\\treturn true;\\n\\t}\\n\\n\\tfunction getBalanceInEth(address addr) public view returns(uint){\\n\\t\\treturn ConvertLib.convert(getBalance(addr),2);\\n\\t}\\n\\n\\tfunction getBalance(address addr) public view returns(uint) {\\n\\t\\treturn balances[addr];\\n\\t}\\n\\n\\tfunction test(ConvertLib.TestStruct memory testObj) public pure {\\n\\t\\trequire(testObj.name > 10);\\n\\t}\\n}\\n\"}},\"settings\":{\"remappings\":[],\"optimizer\":{\"enabled\":true,\"runs\":200},\"evmVersion\":\"berlin\",\"libraries\":{\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/ConvertLib.sol\":{\"ConvertLib\":\"0xd233E7C649ADdCD9752b239013394805f5bE99cc\"}}}}",
  "codeformat": "solidity-standard-json-input",
  "contractname": "/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/MetaCoin.sol:MetaCoin",
  "compilerversion": "v0.8.0+commit.c7dfd78e",
  "constructorArguements": ""
}
Checking status of verification request he3upprdtyarkxjwulsdjsupjzbbnksmtuqxhug7pv3ccdzqip
Pass - Verified: https://rinkeby.etherscan.io/address/0x9731319B3aFE83E2E43e05894eF63A4864897572#contracts
Successfully verified 1 contract(s).
```

Polygon:
```
rosco@rmbp ~/r/e/t/t/metacoin (master) [1]> truffle run verify MetaCoin --network polygon --debug
DEBUG logging is turned ON
Running truffle-plugin-verify v0.5.8
Verifying MetaCoin
Reading artifact file at /Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/build/contracts/MetaCoin.json
Retrieving constructor parameters from https://api.polygonscan.com/api?apiKey=<redacted>&module=account&action=txlist&address=0xE5F4fe685D3c352Fabba76a7999A4D18F0dAF0aD&page=1&sort=asc&offset=1
Constructor parameters retrieved: 0x
Reading artifact file at /Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/build/contracts/ConvertLib.json
Sending verify request with POST arguments:
{
  "apikey": "<redacted>",
  "module": "contract",
  "action": "verifysourcecode",
  "contractaddress": "0xE5F4fe685D3c352Fabba76a7999A4D18F0dAF0aD",
  "sourceCode": "{\"language\":\"Solidity\",\"sources\":{\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/ConvertLib.sol\":{\"content\":\"// SPDX-License-Identifier: MIT\\npragma solidity >=0.4.25 <0.9.0;\\n\\nlibrary ConvertLib {\\n\\tstruct TestStruct {\\n\\t\\tuint256 name;\\n\\t}\\n\\n\\tfunction convert(uint amount,uint conversionRate) public pure returns (uint convertedAmount)\\n\\t{\\n\\t\\treturn amount * conversionRate;\\n\\t}\\n}\\n\"},\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/MetaCoin.sol\":{\"content\":\"// SPDX-License-Identifier: MIT\\npragma experimental ABIEncoderV2;\\npragma solidity >=0.4.25 <0.9.0;\\n\\nimport \\\"./ConvertLib.sol\\\";\\n\\n// This is just a simple example of a coin-like contract.\\n// It is not standards compatible and cannot be expected to talk to other\\n// coin/token contracts. If you want to create a standards-compliant\\n// token, see: https://github.com/ConsenSys/Tokens. Cheers!\\n\\ncontract MetaCoin {\\n\\tmapping (address => uint) balances;\\n\\n\\tevent Transfer(address indexed _from, address indexed _to, uint256 _value);\\n\\n\\tconstructor() public {\\n\\t\\tbalances[tx.origin] = 10000;\\n\\t}\\n\\n\\tfunction sendCoin(address receiver, uint amount) public returns(bool sufficient) {\\n\\t\\tif (balances[msg.sender] < amount) return false;\\n\\t\\tbalances[msg.sender] -= amount;\\n\\t\\tbalances[receiver] += amount;\\n\\t\\temit Transfer(msg.sender, receiver, amount);\\n\\t\\treturn true;\\n\\t}\\n\\n\\tfunction getBalanceInEth(address addr) public view returns(uint){\\n\\t\\treturn ConvertLib.convert(getBalance(addr),2);\\n\\t}\\n\\n\\tfunction getBalance(address addr) public view returns(uint) {\\n\\t\\treturn balances[addr];\\n\\t}\\n\\n\\tfunction test(ConvertLib.TestStruct memory testObj) public pure {\\n\\t\\trequire(testObj.name > 10);\\n\\t}\\n}\\n\"}},\"settings\":{\"remappings\":[],\"optimizer\":{\"enabled\":true,\"runs\":200},\"evmVersion\":\"berlin\",\"libraries\":{\"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/ConvertLib.sol\":{\"ConvertLib\":\"0xeb0B7E59C8037a41dAdb557eBA6684eDea91D016\"}}}}",
  "codeformat": "solidity-standard-json-input",
  "contractname": "/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/MetaCoin.sol:MetaCoin",
  "compilerversion": "v0.8.0+commit.c7dfd78e",
  "constructorArguements": ""
}
Checking status of verification request byxggm4cemeb2edqrpcrrfbsrib7y2dlkvaprfi8cli8nsfxm5
Fail - Unable to verify
Failed to verify 1 contract(s): MetaCoin
```

So the exact same contract code + artifacts can be verified on Rinkeby, but not Polygon. I also tried lower Solidity versions or evm versions to no avail. @Enigmatic331, could you take a look at this issue? I published the WIP polygon support under `truffle-plugin-verify@beta`.

**Comment by hack3r-0m** [comment]
@rkalis can you assign me this issue? i can work on it and submit PR asap

**Comment by rkalis** [comment]
You're welcome to work on it. I have my code in the https://github.com/rkalis/truffle-plugin-verify/tree/polygon branch. The current issue that I'm running into is the one described above, which appears to be on the PolygonScan side. I might be wrong in that, so if you're able to figure out a solution that would be great!

**Comment by hack3r-0m** [comment]
@rkalis I was not able to get it working as well, tried with curl/postman but doesn't seem to work. hardhat-ethers has added support for it https://github.com/nomiclabs/hardhat/commit/273cc92b3dffbab0ba05bc4caa28296c3540ce79, not sure they tested or not

**Comment by rkalis** [comment]
What we could still do is add the Polygon code to the `legacy` branch to see if the verification does work when using flattened source code.

**Comment by hpolthof** [comment]
I've just succesfully verified a few contracts with the polygon branch. Seems that the issues with polygonscan are fixed.

Think it's safe to merge. @rkalis

**Comment by jyap808** [comment]
Can confirm works fine

**Comment by rkalis** [comment]
Nice! Just did a quick check and it's working for me as well.

Released PolygonScan support in v0.5.9 :tada:
